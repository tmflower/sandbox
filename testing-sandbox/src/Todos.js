import { useEffect, useState } from "react";
import axios from "axios";
import { Todo } from "./Todo";

export function Todos() {
    const [todos, setTodos] = useState([]);
    const [message, setMessage] = useState('');
    const [formData, setFormData] = useState({ task: ''});
    const { task } = formData;

    useEffect(() => {
        async function getTodos() {
            const result = await axios.get(`http://localhost:3001/todos`);
            setTodos(result.data)
        }
        getTodos();
    }, [todos]);

    const handleChange = (evt) => {
        const { name, value } = evt.target;
        setFormData(formData => ({ ...formData, [name]: value}));
    }

    const handleSubmit = async (evt) => {
        evt.preventDefault();
        try {
            await axios.post(`http://localhost:3001/todo`, { task });
            setMessage("Task added!");
            setFormData({task: ''})
        }
        catch(err) {
            console.error(err);
        }
    }

    return (
        <div>
            <h1>Todos</h1>
            <h4>Add a new task</h4>
            <form>
                <label htmlFor="task">Task:
                <input 
                    type="text" 
                    name="task" 
                    value={task} 
                    id="task" 
                    onChange={handleChange}>
                </input>
                </label>
                <button onClick={handleSubmit}>Submit</button> 
            </form>
            <p>{message}</p>
            <ul>
                {todos.map((todo, i) => <li key={i} style={{listStyleType:'none'}}><Todo key={i} todo={todo}/></li>)}
            </ul>
        </div>
    )
}