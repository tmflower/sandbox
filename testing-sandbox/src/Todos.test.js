import { render, screen, act } from "@testing-library/react";
import { Todos } from "./Todos";
import userEvent from "@testing-library/user-event";
import axios from "axios";

jest.mock("axios");
const todos = [{ task: "Walk the dog" }, {task: "Feed the fish"}, {task: "Water plants"}];

beforeEach(() => {  
    act(() => {        
        axios.get.mockImplementation(() => Promise.resolve({ data: { todos } }));
      });
});

test("renders without crashing", function() {
    render(<Todos />);
});

test("matches snapshot", () => {
    const { asFragment } = render(<Todos />);
    expect(asFragment()).toMatchSnapshot();
  });

test("renders Todos component correctly", () => {  
    render(<Todos />);      
    const heading = screen.queryByText("Todos");    
    const h4 = screen.getByText("Add a new task");    
    const label = screen.getByText("Task:");    
    const button = screen.getByRole("button");
    screen.debug();
    expect(heading).toBeInTheDocument();
    expect(h4).toBeInTheDocument();
    expect(label).toBeInTheDocument();
    expect(button).toBeInTheDocument();    
});


test("handles submit button click", async() => {
    render(<Todos />);
    const buttonElement = screen.getByRole("button");
    const task = screen.queryByText("Grocery shopping");
    expect(task).not.toBeInTheDocument();

    const newTask = { task: "Grocery shopping" }
    await axios.post.mockImplementation(() => Promise.resolve({ data: { newTask } }));
    await userEvent.click(buttonElement);

    const newTaskRendered = screen.getByText("Grocery shopping");
    expect(newTaskRendered).toBeInTheDocument();
});