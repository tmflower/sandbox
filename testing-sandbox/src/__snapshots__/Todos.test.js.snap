// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Todos
    </h1>
    <h4>
      Add a new task
    </h4>
    <form>
      <label
        for="task"
      >
        Task:
        <input
          id="task"
          name="task"
          type="text"
          value=""
        />
      </label>
      <button>
        Submit
      </button>
    </form>
    <p />
    <ul />
  </div>
</DocumentFragment>
`;
